backend/worker/kafka_worker.go

```diff
@@ -121,30 +124,113 @@ func (k *KafkaBatchWorker) flush(ctx context.Context) {
+			log.WithContext(ctxTs).WithError(err).Errorf("failed to set last log timestamp for project %d", projectId)
+		}
+	}
+	spanTs.Finish()
+
+	spanW, ctxW := tracer.StartSpanFromContext(ctx, "kafkaBatchWorker", tracer.ResourceName("worker.kafka.batched.checkBillingQuotas"))
+
+	// If it's saved in Redis that a project has exceeded / not exceeded
+	// its quota, use that value. Else, add the projectId to a list of
+	// projects to query.
+	quotaExceededByProject := map[uint32]bool{}
+	projectsToQuery := []uint32{}
+	for projectId := range workspaceByProject {
+		exceeded, err := k.Worker.Resolver.Redis.IsBillingQuotaExceeded(ctxW, int(projectId), pricing.ProductTypeLogs)
+		if err != nil {
+			log.WithContext(ctxW).Error(err)
+			continue
+		}
+		if exceeded != nil {
+			quotaExceededByProject[projectId] = *exceeded
+		} else {
+			projectsToQuery = append(projectsToQuery, projectId)
+		}
+	}
+
+	// For any projects to query, get the associated workspace,
+	// check if that workspace is within the logs quota,
+	// and write the result to redis.
+	for _, projectId := range projectsToQuery {
+		var project model.Project
+		if err := k.Worker.Resolver.DB.Model(&project).
+			Where("id = ?", projectId).Find(&project).Error; err != nil {
+			log.WithContext(ctxW).Error(e.Wrap(err, "error querying project"))
+			continue
+		}
+
+		var workspace model.Workspace
+		if err := k.Worker.Resolver.DB.Model(&workspace).
+			Where("id = ?", project.WorkspaceID).Find(&workspace).Error; err != nil {
+			log.WithContext(ctxW).Error(e.Wrap(err, "error querying workspace"))
+			continue
+		}
+
+		projects := []model.Project{}
+		if err := k.Worker.Resolver.DB.Order("name ASC").Model(&workspace).Association("Projects").Find(&projects); err != nil {
+			log.WithContext(ctxW).Error(e.Wrap(err, "error querying associated projects"))
+			continue
+		}
+		workspace.Projects = projects
+
+		withinBillingQuota, quotaPercent := k.Worker.PublicResolver.IsWithinQuota(ctxW, pricing.ProductTypeLogs, &workspace, time.Now())
+		quotaExceededByProject[projectId] = !withinBillingQuota
+		if err := k.Worker.Resolver.Redis.SetBillingQuotaExceeded(ctxW, int(projectId), pricing.ProductTypeLogs, !withinBillingQuota); err != nil {
+			log.WithContext(ctxW).Error(err)
+		}
+
+		// Send alert emails if above the relevant thresholds
+		go func() {
+			defer util.Recover()
+			if quotaPercent >= 1 {
+				if err := model.SendBillingNotifications(ctx, k.Worker.PublicResolver.DB, k.Worker.PublicResolver.MailClient, email.BillingLogsUsage100Percent, &workspace); err != nil {
+					log.WithContext(ctx).Error(e.Wrap(err, "failed to send billing notifications"))
+				}
+			} else if quotaPercent >= .8 {
+				if err := model.SendBillingNotifications(ctx, k.Worker.PublicResolver.DB, k.Worker.PublicResolver.MailClient, email.BillingLogsUsage80Percent, &workspace); err != nil {
+					log.WithContext(ctx).Error(e.Wrap(err, "failed to send billing notifications"))
+				}
+			}
+		}()
```

Comment:
nit: these billing notification inlined fns could be their own package, esp since we have them in a few different places in the ingest path (sessions, errors, logs). `quotaPercent` could be a const?

Reply:
:+1: will do this in a follow up

Reply:
sgtm